`qemu-utils`: reduce build closure size

roberth opened the pull request:
## Description of changes

Like #251046, but an extra step smaller by omitting many build dependencies from the `qemu` input.

- qemu-utils: Rename parameter to qemu
  
  Allows overriding without prior knowledge of which qemu package attribute name is the default.

- qemu-utils: Use smaller qemu_test as input

  1823 -> 990 derivations in the build closure.

  This saves rebuilds and makes qemu-utils less susceptible to build errors.

  qemu_test and qemu_utils both need network- and disk-related functionality, so they should be a good match.


**Alternative I rejected**

I have considered an alternate solution to my problem. The NixOS test framework would tell `make-disk-image.nix` to use its `virtualisation.qemu`, so that `qemu-utils` isn't needed at all, but this would cause images to be rebuilt when switching to the interactive driver, which uses a fancier `qemu` with graphical dependencies, etc. I consider this to be too magical and a place for heisenbugs to hide. [Code](https://github.com/NixOS/nixpkgs/compare/master...hercules-ci:nixpkgs:nixos-test-disk-image-reuse-qemu) for posterity.

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
  - Tested with `nixosTests.gitlab` after setting `virtualisation.writableStore = true;` so that `qemu-utils` is used.
  - `nix-store -qR $(nix-instantiate -A nixosTests.gitlab) | grep gtk` (`gtk-doc` is expected; `gtk+` would not be)
- [23.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) (or backporting [23.05 Release notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md))
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

roberth committed:
qemu-utils: Rework as an emulatorless qemu build

Benefits

  - Reduce the derivation closure.
      - This reduces rebuilds and makes derivations that use
        qemu-utils less susceptible to build errors.
  - Includes extra qemu tools that weren't included before.
  - Removes fragile binary copying (old qemu-utils impl).
      - Can't forget to copy somethat that a binary needs.
      - Less custom packaging is better.

Cost

  - Adds an 80 second build, but the time is offset by rebuilding
    its dependents less.

  - The output of qemu-utils goes from ~7 MB to ~40 MB, but it
    provides more functionality.

alyssais commented:
I'm not sure about this — it has those positive effects now, but one of the things that makes qemu_test qemu_test is that it's patched in ways that are specific to the NixOS test runner.  I don't think the one patch we currently have would affect qemu-utils, but I'm concerned that in future that might not be the case.

cole-h commented:
Is the patch what makes qemu_test small? If not, I wonder if it would make sense to add a qemu_super_minimal or something that has the same small closure size without NixOS specific patch(es), and build qemu_test on top of that...?

alyssais commented:
Then we end up with yet another QEMU build.  I'm not sure that's an improvement.  (There's also the secret QEMU build defined in lib (eww) used for emulation.)

roberth commented:
Found a better way. An empty list of architectures can be specified, in which case it produces mostly just the tools, in about 80 seconds.

<details>

Without extraction (ie just overrides; internal package without attribute (for now?))

```console
$ ls result/
bin/     include/ libexec/ share/

$ ls result/bin/
elf2dmp  qemu-edid  qemu-img  qemu-io  qemu-kvm  qemu-nbd  qemu-pr-helper  qemu-storage-daemon

$ nix-store -q --references result/
/nix/store/ld03l52xq2ssn4x0g5asypsxqls40497-glibc-2.37-8
/nix/store/ig0kkzw4n2pws12dj7szjm71f1a43if6-zlib-1.3
/nix/store/1c6x8n47mq5rcivsp72b0dyqgjghi6yx-glib-2.76.4
/nix/store/3b3mzw23bi8ga9jg8lh2pp0j61qnib77-liburing-2.4
/nix/store/5p7jmbpmq7clb3whbi6kyp50gmg54l2d-zstd-1.5.5
/nix/store/axyr8b6dmv24vxric98248hrb3nqnf7l-libaio-0.3.113
/nix/store/bhmynyjwzc2r6iqf7fhc3yzjcv3paiwa-curl-8.3.0
/nix/store/vawph0vsv3zwvknk6rhml51lx33lcspd-libcap-ng-0.8.3
/nix/store/646d9mid09n8h2zl8196l7912hziw92n-qemu-8.1.2

$ nix-store -q --size result
41949776
```

With extraction (ie `qemu-utils` proper)

```console
$ ls result/
bin  share

$ ls result/bin/
qemu-img  qemu-io  qemu-nbd

$ nix-store -q --size result
7361824

$ nix-store -q --references result/
/nix/store/ld03l52xq2ssn4x0g5asypsxqls40497-glibc-2.37-8
/nix/store/ig0kkzw4n2pws12dj7szjm71f1a43if6-zlib-1.3
/nix/store/1c6x8n47mq5rcivsp72b0dyqgjghi6yx-glib-2.76.4
/nix/store/3b3mzw23bi8ga9jg8lh2pp0j61qnib77-liburing-2.4
/nix/store/5p7jmbpmq7clb3whbi6kyp50gmg54l2d-zstd-1.5.5
/nix/store/axyr8b6dmv24vxric98248hrb3nqnf7l-libaio-0.3.113
/nix/store/bhmynyjwzc2r6iqf7fhc3yzjcv3paiwa-curl-8.3.0
```

</details>

alyssais commented on pkgs/top-level/all-packages.nix:
```diff
@@ -34933,7 +34933,12 @@ with pkgs;
+  qemu-utils = callPackage ../applications/virtualization/qemu/utils.nix {
+    qemu = qemu.override (o: {
+      toolsOnly = true;
+      hostCpuTargets = [ ];
```
Shouldn't this be implied by `toolsOnly`?

roberth replied:
Done.

alyssais commented on pkgs/top-level/all-packages.nix:
```diff
@@ -34933,7 +34933,12 @@ with pkgs;
   };
 
-  qemu-utils = callPackage ../applications/virtualization/qemu/utils.nix { };
+  qemu-utils = callPackage ../applications/virtualization/qemu/utils.nix {
```
Why do we still need utils.nix?  How much space are we saving?

roberth replied:
Some 35 MB on top of the 7 MB we had. It appears to be more useful and correct this way though.

roberth committed:
qemu: Build qemu-utils on ofborg

Makes sure that the qemu-utils build succeeds when qemu-utils is
about to be updated. It's a bit of an odd build perhaps, so it's
good to check that it still builds.

alyssais commented on pkgs/applications/virtualization/qemu/default.nix:
```diff
@@ -243,10 +248,62 @@ stdenv.mkDerivation (finalAttrs: {
+
+  installCheckPhase =
+    if toolsOnly
+    then ''
```
lib.optionalString

roberth replied:
`null` avoided a rebuild of `qemu`, but I guess `optionalString` (`""`) is more idiomatic.

alyssais commented on pkgs/applications/virtualization/qemu/default.nix:
```diff
@@ -243,10 +248,62 @@ stdenv.mkDerivation (finalAttrs: {
+              if test -L $bin && ! test -e $bin; then
+                echo "removing broken $bin symlink"
+                echo "upstream build system should not write it, but it does, and it's not a big deal"
+                rm $bin
```
This seems like a very weird thing to be doing in installCheckPhase.

roberth replied:
I made a mistake. Thanks!

alyssais commented on pkgs/applications/virtualization/qemu/default.nix:
```diff
@@ -243,10 +248,62 @@ stdenv.mkDerivation (finalAttrs: {
+            qemu-kvm)
+              if test -L $bin && ! test -e $bin; then
+                echo "removing broken $bin symlink"
+                echo "upstream build system should not write it, but it does, and it's not a big deal"
```
Can we fix this?

roberth replied:
Done.

alyssais commented on pkgs/applications/virtualization/qemu/default.nix:
```diff
@@ -243,10 +248,62 @@ stdenv.mkDerivation (finalAttrs: {
+            qemu-nbd) ;;
+            qemu-pr-helper) ;;
+            qemu-storage-daemon) ;;
+            qemu-kvm)
```
qemu-kvm is not something the upstream build system generates.

alyssais commented on pkgs/applications/virtualization/qemu/default.nix:
```diff
@@ -243,10 +248,62 @@ stdenv.mkDerivation (finalAttrs: {
+              echo "unexpected binary in qemu-utils package: $bin"
+              echo "qemu-utils is not supposed to contain the emulator"
+              exit 1
+              ;;
```
This seems overly defensive to me.  We're explicitly telling it not to build any system emulators.  I don't think we need to second guess the build system.

roberth replied:
Removed the special case, but not the entire check. I think it's good to control which binaries are in the package.

alyssais replied:
I'm not sure — given that we're happy including everything that's currently there, I think it's very likely we'll allow everything in the future anyway.

roberth replied:
Ok. I've removed the check.

alyssais commented on pkgs/top-level/all-packages.nix:
```diff
@@ -34933,7 +34933,9 @@ with pkgs;
   };
 
-  qemu-utils = callPackage ../applications/virtualization/qemu/utils.nix { };
+  qemu-utils = qemu.override (o: {
```
`override` can take an attrset — no need for a function.

alyssais reviewed (approved):
LGTM with a squash.

alyssais commented:
BTW a future enhancement possibly worth considering would be _only_ building these tools in qemu-utils, and then symlinking to them from the other qemu packages, so we'd only build them once, and deduplicate them.

roberth commented:
If the configuration flags don't affect the utils too much, that should work.
Even if some flags need to be propagated, the qemu-utils derivation might still converge to like two variations or something.
